From 235f72897aa48a47066e40d8ade1693659627783 Mon Sep 17 00:00:00 2001
From: Szabolcs Nagy <nsz@port70.net>
Date: Sun, 21 Jun 2026 11:33:28 +0000
Subject: wordexp: fix stderr redirection

stderr redirection to /dev/null didn't work, can be done in the
command, but would not redirect errors printed during sh startup.

  wordexp(")", p, 0)

clobbered stderr, now silent.
---
 src/misc/wordexp.c | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/src/misc/wordexp.c b/src/misc/wordexp.c
index db3e0067..5f9bc1d5 100644
--- a/src/misc/wordexp.c
+++ b/src/misc/wordexp.c
@@ -29,8 +29,8 @@ static int do_wordexp(const char *s, wordexp_t *we, int flags)
 	int sq=0, dq=0;
 	size_t np=0;
 	char *w, **tmp;
-	char *redir = (flags & WRDE_SHOWERR) ? "" : "2>/dev/null";
 	int err = 0;
+	int null = -1;
 	FILE *f;
 	size_t wc = 0;
 	char **wv = 0;
@@ -97,23 +97,31 @@ static int do_wordexp(const char *s, wordexp_t *we, int flags)
 	}
 
 	if (pipe2(p, O_CLOEXEC) < 0) goto nospace;
+	if (!(flags & WRDE_SHOWERR)) {
+		null = open("/dev/null", O_WRONLY|O_CLOEXEC);
+		if (null < 0) goto nospace;
+	}
 	__block_all_sigs(&set);
 	pid = fork();
 	__restore_sigs(&set);
 	if (pid < 0) {
 		close(p[0]);
 		close(p[1]);
+		close(null);
 		goto nospace;
 	}
 	if (!pid) {
 		if (p[1] == 1) fcntl(1, F_SETFD, 0);
 		else dup2(p[1], 1);
+		if (null == 2) fcntl(2, F_SETFD, 0);
+		else dup2(null, 2);
 		execl("/bin/sh", "sh", "-c",
-			"eval \"printf %s\\\\\\\\0 x $1 $2\"",
-			"sh", s, redir, (char *)0);
+			"eval \"printf %s\\\\\\\\0 x $1\"",
+			"sh", s, (char *)0);
 		_exit(1);
 	}
 	close(p[1]);
+	close(null);
 	
 	f = fdopen(p[0], "r");
 	if (!f) {
-- 
cgit v1.2.1

